### ISSUE TITLE
How to supervised finetuning mask in torchtitan?

### ISSUE THREAD

**Issue by kailashg26** [pytorch/torchtitan#1781]
How do I implement supervised fine-tuning (SFT) masking in TorchTitan for posttraining using a synthetic dataset?

**Comment by kailashg26** [comment]
@tianyu-l , is there a PR on going for supporting SFT mask in torchtitan? or is this a different issue?
https://github.com/pytorch/torchtitan/pull/1321

**Comment by tianyu-l** [comment]
There is no ongoing effort right now in torchtitan. I think the main work here is to support
1. a new dataset example
2. masking in loss computation

IIUC 2 should be shared with VLM training as well.
cc @wwwjn @lkhphuc

**Comment by kailashg26** [comment]
@tianyu-l Thanks for quick response. If I understand correctly, the above PR is using C4 dataset to finetune the pretrained llama 3 8B model? C4 is basically the same dataset used for pre-training as well right? Please lmk if I'm missing something :)
Also, Could you please share some insights into how to do SFT in torchtitan? Would be really helpful.

**Comment by kailashg26** [comment]
@wwwjn @lkhphuc @tianyu-l Could you please provide some insights into this?

**Comment by tianyu-l** [comment]
I think the main missing piece is data loading for SFT.

For examples of training on a fine-tuning dataset, you could refer to how torchtune handles SFT data loading, e.g. https://github.com/meta-pytorch/torchtune/blob/main/torchtune/datasets/_sft.py

If we could find bandwidth, we may consider adding an example in torchtitan.

**Comment by kailashg26** [comment]
I can actually try adding one for torchtitan. But I'm more familiar with torchtune. So, if you could please help with steps or something like that I can start working on it :) Thanks @tianyu-l

**Comment by tianyu-l** [comment]
Oh, is it hard to switch torchtitan dataloader to a torchtune one? Curious what difficulties you hit?

**Comment by kailashg26** [comment]
My bad, should have explained better. Since torchtune has stopped dev tasks. I was curious to try finetuning on torchtitan. So, I was wondering if we can do finetuning on torchtitan. From my understanding in order to try SFT on torchtitan, I need to change the data loader, loss masking, and the dataset should be prompt-response right? Please let me know if I'm missing any other things.

**Comment by tianyu-l** [comment]
> change the data loader, loss masking, and the dataset should be prompt-response right?

That sounds right to me.

Another piece which may be involved is tokenizer, as you might need some special tokens for specific tasks, e.g. Q&A, reasoning.

**Comment by JenniferWang** [comment]
@kailashg26 , @joecummings is fixing the SFT recipe from the torchforge side. Since torchforge uses Titan as the trainer backend, we noticed the similar problem https://github.com/meta-pytorch/torchforge/pull/614.

Shall we combine the effort and propose a solution for SFT with Titan?

**Comment by rakkit** [comment]
@tianyu-l I was checking around the SFT receipt from [verl](https://github.com/volcengine/verl/blob/main/verl/workers/utils/losses.py#L27) and [openrlhf](https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/trainer/sft_trainer.py#L152)
To make SFT work smoothly, we have to pass extra args (like mask) to `loss_fn`.  It's kind of incompatible with current torchtitna's `loss_fn` (read only two tensors), any suggestion about that?

**Comment by lkhphuc** [comment]
For [CrossEntropyLoss](https://docs.pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), you can simply make the dataloader return index=-100 for whichever tokens need to be masked out from the loss computation. 
This correspond to the default `ignore_index` of pytorch's CE loss.
`labels[start_of_question_idx:start_of_answer]=100` something like this.
In essence, the only difference between SFT and pretraining is only in the dataloader.

**Comment by rakkit** [comment]
Thanks all, I opened PR here https://github.com/pytorch/torchtitan/pull/2148 for SFT training, it will be good if you can help review it and give any feedback
